### ISSUE TITLE
Erase in Line ANSI code `\x1B[K` overwrites last character in full-width line

### ISSUE THREAD

**Issue by cdelledonne** [ninja-build/ninja#2209]
In "smart terminals", Ninja uses the ANSI code `\x1B[K` (EL) to erase from the cursor to the end of line, so as to clean whatever was left from the previous line of output
https://github.com/ninja-build/ninja/blob/91f99ab60747e76eca5222c74531268a5d55acf1/src/line_printer.cc#L109-L115

When the current line of output is just as long as the width of the terminal (which happens with all elided lines), the last character of the line gets overwritten by the EL code. This behavior actually depends on the implementation of the terminal. However, as hinted in this related issue https://github.com/alacritty/alacritty/issues/6159, this seems like the intended behavior for terminals that comply with the standard mentioned in that issue. For instance, this command should result in a `>` character at the end of the first line, but such character is overwritten if the command is run in e.g. XTerm
```bash
printf '-%.0s' {1..$(($COLUMNS-1))} ; printf '>' ; printf '\e[K' ; printf '\r\nSecond line\n'
```

Apparently, according to the standard mentioned in the linked issue, when a character is printed on the last column of the terminal (`>` in the command above), the column index is not incremented, and thus the EL code ends up overwriting that last character.

Could a solution be implemented in Ninja? For instance, send the EL code at the beginning of the line as opposed to as the end
```cpp
winsize size;
if ((ioctl(STDOUT_FILENO, TIOCGWINSZ, &size) == 0) && size.ws_col) {
  to_print = ElideMiddle(to_print, size.ws_col);
}
printf("\x1B[K");  // Clear to end of line.
printf("%s", to_print.c_str());
fflush(stdout);
```
or only send the EL code if the line is shorter than the width of the terminal
```cpp
winsize size;
if ((ioctl(STDOUT_FILENO, TIOCGWINSZ, &size) == 0) && size.ws_col) {
  to_print = ElideMiddle(to_print, size.ws_col);
}
printf("%s", to_print.c_str());
if (to_print.size() < size.ws_col) {
  printf("\x1B[K");  // Clear to end of line.
}
fflush(stdout);
```

**Comment by cdelledonne** [comment]
The macOS Terminal seems to suffer from this issue. Other terminal emulators where this issue seems to appear include Alacritty, the terminal embedded in Neovim (VT220/xterm terminal emulator) and the terminal embedded in Vim. And possibly some other ones which I haven't tested.

I'm not sure what adjusting the standard would entail. I also don't know if the terminal emulators would adopt any such changes promptly.

Out of curiosity, would you anticipate any problems with Ninja implementing an application-level solution instead?

**Comment by jhasse** [comment]
> The macOS Terminal seems to suffer from this issue.

Just tried it myself and you're right. iTerm2 doesn't though, at least.

> I'm not sure what adjusting the standard would entail.

We could tell those terminals "You're wrong" ;)

> Out of curiosity, would you anticipate any problems with Ninja implementing an application-level solution instead?

Your second solution looks reasonable. I still think that GCC's output is a lot more important than Ninja's status line, so I don't think fixing this in Ninja needs to come before GCC has figured out a way forward.

**Comment by cdelledonne** [comment]
I agree that GCC's output is a lot more important than Ninja's status line, however they are two independent things. GCC could potentially end up addressing this in years, seeing how old the bug report already is.

Anyway, I don't think this is a high-priority issue, as it does not affect Ninja's functionality.
